Instructor can add no of characters limit on the field

ravimajithia opened the pull request:
- No of characters allowed for field from editor
- Improvements in validation design
- Indicator for remaning characters in field for readers

Commit:
Instructor can add no of characters limit on the field
- No of characters allowed for field from editor
- Improvements in validation design
- Indicator for remaning characters in field for readers

otacke commented on semantics.json:
```diff
@@ -20,6 +20,14 @@
     "description": "Learners must provide an answer in order to proceed"
   },
+  {
+    "label": "Characters Limit",
```
I think the unspoken convention is to only capitalize the first letter of the first word, so "Character limit" maybe?

otacke commented on semantics.json:
```diff
@@ -20,6 +20,14 @@
+    "type": "number",
+    "importance": "high",
+    "default": "200",
+    "description": "Number of characters allowed in the field"
```
Normally, the description fields end with a "." even if they do not contain a full sentence.

otacke commented on semantics.json:
```diff
@@ -20,6 +20,14 @@
+    "name": "charactersLimit",
+    "type": "number",
+    "importance": "high",
+    "default": "200",
```
Should there be a default set? I don't know. I think @janlindso may need to decide that.

If it's set, it should be a `number`, not a `string` though (cmp. https://h5p.org/semantics#type-number). Interesting that the `number` editor widget even keeps the string and doesn't parse it into a number.

ravimajithia replied:
Yes, I think it is needed because we allow the user to change this limit from the editor, and if the user decided to submit an empty field, that will break the calculation logic. I am still open to suggestions!

otacke replied:
Not sure if I understand this correctly: I was under the assumption that the character limit was optional. If it is, then a default value should be set when instantiating the content type anyway. All non-optional values need to be sanitized if they are required. One cannot assume that all parameters are set. For instance, when subcontent is used in Course Presentation, the visual editor will instantiate subcontent with empty parameters - this must not crash content.

ravimajithia replied:
In my mind, I was thinking that we want users to add content but, not too much. So I approach this task with some upper limit in the number of characters. That is the reason it is required and high importance.

otacke replied:
Guess we'll have to wait for some feedback from @janlindso or Knut-Inge then. I don't know if a limit should be mandatory. If it is, then yes, importance should be "high".

ravimajithia replied:
It's been updated and revised.

otacke commented on semantics.json:
```diff
@@ -20,6 +20,14 @@
+    "label": "Characters Limit",
+    "name": "charactersLimit",
+    "type": "number",
+    "importance": "high",
```
Is "high" appropriate here? I never questioned the importance of the `importance` field. What is the rule at H5P Group?

ravimajithia replied:
I am not sure about the rule but the document suggests this - 
"An indication of the field's importance. Is typically used by the editor giving a more prominent style on the more important fields, but could e.g. also be used to generate weighted input to a search engine."

I am open to suggestions but I do not see any issue with high importance here.

otacke replied:
See my comment on the character limit field being optional: If it is optional, than the importance definitely is not high. If it is, it is.

ravimajithia replied:
We can merge this in the previous conversation. Hence, resolving this conversation.

otacke commented on semantics.json:
```diff
@@ -59,6 +67,14 @@
+        "name": "remainingChars",
+        "type": "text",
+        "label": "Remaining characters",
+        "importance": "low",
```
I think `low` is implied automatically, so it could be spared (goes for the other translation fields, too).

ravimajithia replied:
Just maintaining consistency with other fields.

otacke replied:
:-D Then all the other fields should be changed IMHO, but it's really no matter of life and death.

otacke commented on src/scripts/components/elements/status-bar.js:
```diff
@@ -0,0 +1,45 @@
+   * @class
+   * @param {object} params Parameter from editor.
+   * @param {object} params.i10n Localization strings.
+   * @param {object} [callbacks] Callbacks.
```
Would not hurt to have `@param {function} [callbacks.onClick] Callback for click.` here as well.

ravimajithia replied:
There is no need for a callback, I am not using this in any manner and I shouldn't have included this in the code for this file. I will remove this from the code.

otacke replied:
You got me here :-D I should have seen this.

ravimajithia replied:
I should've seen this before committing, it's not the reviewer's responsibility but the dev's 😄

otacke replied:
Nah, if I check things, I should check what's actually needed, too. Responsibility for the quality in my humble opinion is shared among everyone who contributes to the code in some way - so there's no point in "blaming" anyone.

otacke commented on src/scripts/components/elements/status-bar.js:
```diff
@@ -0,0 +1,45 @@
+    // Statusbar
+    this.dom = document.createElement('div');
+    this.dom.classList.add('h5p-reader-question-text-limit');
+    const remainingChars = this.params.charactersLimit - this.params.initialChars;
```
Isn't this what `setUpdatedCharsCount` should be doing? Could be called here. ("Don't repeat yourself")

ravimajithia replied:
No, it does not calculate the remaining characters but the rest is handled by it. I will update the code.

otacke replied:
No, of course it doesn't compute that value, but it has this nice argument that would take it: `setUpdatedCharsCount(this.params.charactersLimit - this.params.initialChars);` and you'd not do the same thing in two places.

ravimajithia replied:
Thank you for the suggestion, and as mentioned in my previous comment I will update the code.

otacke commented on src/scripts/components/elements/status-bar.js:
```diff
@@ -0,0 +1,45 @@
+    if (remainingChars < 0) {
+      this.dom.classList.add('error-exceeded-chars');
+    }
+    this.dom.innerHTML = this.params.i10n.remainingChars.replace(/@chars/g, remainingChars);
```
This is not a hard rule, but please try to keep a line within 80 characters and break it appropriately if possible.

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
   }
+
+  /**
+   * Validate input/editor content
```
Try to keep the JSDoc comments uniform
- with a "." at the end of each line and
- starting the description text with a capital letter.

ravimajithia replied:
Got it! FYI it is in capital.

otacke replied:
"length of content" is not. Sorry for not adding a comment to every single line.

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
+  /**
+   * Validate input/editor content
+   * @param {number} contentLength length of content
+   * @returns {boolean}.
```
It's good practice to add a description, here e.g. `True, if input validates. Else false.`, so if one created automated documentation from JSDoc, the values would bear meaning.

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
+  validate(contentLength) {
+    const remainingChars = this.params.charactersLimit - contentLength;
+    this.params.fieldContainer.classList.remove('validation-error');
+    let isError = this.validateCharsLimit(contentLength) ? true : false;
```
If the return value of `validateCharsLimit` is a `boolean` (it is), then using `let isError = this.validateCharsLimit(contentLength);` is sufficient. One does not need to explicitly return `true` if the return value is `true` and `false` if the return value is `false`.

If you think that you'd like this to be more "speaking", you can rename `validateCharsLimit` to something like `isCharLimitExceeded`. Then you'd have `let isError = this.isCharLimitExceeded(contentLength);` which should then be pretty self-explanatory and shorter.

ravimajithia replied:
Yes, that makes sense! Thanks.

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
+      isError = true;
+    }
+
+    return isError;
```
I am not 100% sure, but I think in H5P core `validate` usually returns `true` if everything is fine and `false` in case of an error. One could rename `validate` as well, however.

ravimajithia replied:
Hmm... I am not really following H5P rules for the validation but not sure if that is necessary in this case as well.

For simplicity and to avoid confusion I will rename the function.

otacke replied:
It's not necessary for us, no. But I am trying to follow the premise that even less experienced people should not have to struggle with H5P code too much. If `validate` always follows the same scheme in every corner of H5P, that's probably better - I know, just a tiny drop in the ocean, but still a drop.

I hope my comments do not come across as "This must be the way!!!".

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
+  }
+
+  /**
+   * Validate input/editor content
```
See above comments about JSDoc.

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
+   */
+  validateCharsLimit(contentLength) {
+    const remainingChars = this.params.charactersLimit - contentLength;
+    let isError = false;
```
The variable `isError` is no really required here if there's just one condition it depends on. You can simple return `true` inside the comparison of `remainingChars < 0` and `false` afterwards.

otacke commented on src/scripts/components/elements/validation.js:
```diff
@@ -66,4 +73,49 @@ export default class Validation {
+  }
+
+  /**
+   * Updated field container parameter
```
See comments about JSDoc.

otacke commented on src/scripts/components/main.js:
```diff
@@ -144,6 +188,37 @@ export default class Main {
+   */
+  handleTextChanged() {
+    const content = this.getPlaintextContent();
+    const remainingChars = parseInt(this.globalParams.charactersLimit) - content.length;
```
I think you will always want `charactersLimit` to be a number, so you should sanitize it when instantiating `ActiveReaderTextInput`right away. In turn, retrieving the number of remaining characters will be a simple computation that could be the argument of the next line directly.

ravimajithia replied:
It's been taken care of

otacke commented on src/scripts/components/main.js:
```diff
@@ -144,6 +188,37 @@ export default class Main {
+  getPlaintextContent() {
+    const tempEle = document.createElement('div');
+    tempEle.innerHTML = this.getResponse();
+    const content = tempEle.innerText || tempEle.textContent;
```
No need to have this `content` variable here. You can return the result directly.

Also, this would be a typical "stripHTML" function that could go into `Util`, e.g. as:
```
  /**
   * Strip HTML.
   * @param {string} html HTML string.
   * @returns {string} Plain text.
   */
  static stripHTML(html) {
    if (typeof html !== 'string') {
      return '';
    }

    const div = document.createElement('div');
    div.innerHTML = html;
    return div.textContent || div.innerText || '';
  }
```

Then, `getPlaintextContent` would only be `return Util.stripHTML(this.getResponse());`

otacke commented on src/scripts/components/elements/status-bar.js:
```diff
@@ -0,0 +1,45 @@
+   * @param {number} remainingChars for this class.
+   */
+  setUpdatedCharsCount(remainingChars) {
+    this.dom.classList.remove('error-exceeded-chars');
```
Not sure if this is really needed, but essentially
`this.dom.classList.toggle('error-exceeded-chars', remainingChars < 0)` would be sufficient to handle both, removal and addition.

otacke reviewed (changes requested):
Hey! Thanks a lot. I have a couple of small suggestions.

I think the positioning of the status bar feels a little off. There's quite a large gap underneath the text input field (because of its padding) and hardly any space to the submit button (on a narrow view). Maybe @janlindso has some preference here.

The `resetTask` function will also need to take care of the status bar now.

Commit:
Improve code
- Resolve minor issues
- Improve naming conventions and make eslint happy
- Remove min number of characters cap from the text field
- Remove unwanted code

Commit:
Improve code
- Add a11y support for exceeding characters limit
- Design improvements
- Resolve bugs onreset task

Commit:
Make JSDoc happy

ravimajithia commented:
> https://github.com/NDLANO/h5p-active-reader-text-input/pull/3#pullrequestreview-1420047101

The `resetTask` function is revised and working as expected.
Regarding the design changes, I have updated the CSS to look more appropriate. Please have a look.
